### DISCUSSION TITLE
Sendgrid credentials, connections and deprecation

### DISCUSSION THREAD

**Discussion by ggsdc** [apache/airflow#34803] [Q&A] [answered]
Hi!

Following the [tutorial](https://airflow.apache.org/docs/apache-airflow/stable/howto/email-config.html#using-sendgrid-provider) for SendGrid configuration I have created the configuration via the airflow UI > admin > conenctions and it works properly.

But I would like to not have to create the connection manually, I tried creating the connection as it proposes in the managing connections [tutorial](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html) but I have not been able to make it work and I get a deprecation warning that email credential can not be read from [environment variables](https://github.com/apache/airflow/pull/13601).

I tried to create the connection as a [URI](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#uri-format-example), but it did not work.

So the questions would be two:

1. How can I create the connection without using the user interface when deploying airflow (for the first time or because I need to change the password). This either could be done with the environment file for a docker compose set up or in the Dockerfile? But I guess that I will get the deprecation warning right?
2. Is the deprecation going to take place at any point in time or specific version?

Thank you in advance

EDIT: a little more context.

If I get the connection URI with `airflow connections get sendgrid_conn` (that was the name that I gave it) and create a environment variable with the name `AIRFLOW_CONN_SENDGRID_CONN` it fails to send the email with a `401 Unathorized` error but it will not raise the deprecation warning.

The URI has the following structure: `email://apikey:{PASSWORD}@smpt.sendgrid.net:587`.

Any clues to why is this happening?

**Comment by Taragolis** [comment, marked as answer]
> but I have not been able to make it work and I get a deprecation warning that email credential can not be read from https://github.com/apache/airflow/pull/13601.

Without logs it is quite difficult to say. I think that this deprecation only about plain credentials into the config, you should still able to use regular connection. And it show every-time if connection id not exists

> How can I create the connection without using the user interface when deploying airflow

Airflow read connections and variables by this order:
1. Check in Alternative Secret Backend (if configured)
2. Check in environment variables
3. Airflow DB

So in your case if you need to create conn_id before Airflow started you need to use Secrets Backend or Environment variables. Please note: Connection in Secrets Backend and Env Vars doesn't shown into the Airflow UI and into the cli

Here the simple snippet for generate connection URI / JSON which could be used:
```python
#  You could run it in python venv with preinstalled airflow just as regular python module
import os
import json
from airflow.models.connection import Connection
from airflow.hooks.base import BaseHook

conn = Connection(
    conn_id="sample_sendgrid_connection",
    conn_type="email",
    password="I-Really-Dont-Know-Format-Of-Sendgrid-Key",
)

# Option 1: Generate Environment Variable Name and Connection URI
env_key = f"AIRFLOW_CONN_{conn.conn_id.upper()}"
conn_uri = conn.get_uri()
print(f"{env_key}={conn_uri}")
# AIRFLOW_CONN_SAMPLE_SENDGRID_CONNECTION=email://:I-Really-Dont-Know-Format-Of-Sendgrid-Key@

# Just for check
os.environ[env_key] = conn_uri
# Should not raise any error
conn_from_envars = BaseHook.get_connection("sample_sendgrid_connection")
assert conn_from_envars.password == "I-Really-Dont-Know-Format-Of-Sendgrid-Key"


# Option 2: Generate Environment Variable Name and Connection JSON
env_key = f"AIRFLOW_CONN_{conn.conn_id.upper()}"
conn_json = json.dumps({"conn_type": "email", "password": "Foo-Bar-Spam-Egg"})

print(f"{env_key}='{conn_json}'")
# AIRFLOW_CONN_SAMPLE_SENDGRID_CONNECTION='{"conn_type": "email", "password": "Foo-Bar-Spam-Egg"}'

# Just for check
os.environ[env_key] = conn_json
# Should not raise any error
conn_from_envars = BaseHook.get_connection("sample_sendgrid_connection")
assert conn_from_envars.password == "Foo-Bar-Spam-Egg"
```

So in this case you need to:
1. Create connection suitable for Env Var (URI/JSON)
2. Provide into your containers `AIRFLOW_CONN_UPPER_CASE_OF_CONN_ID`
3. Set https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#email-conn-id by your conn id, e.g. `sample_sengrid_connection`

**Reply by ggsdc** [reply]
Thank you for your help.

The main problem I was having was that if you do `airflow connections get NAME_OF_CONNECTION` it will give you back a connection URI which was not correct due to the password being a wrong one (is similar but not the same). Is this intended? Is this an issue?

Once I found out that I was able to make it work creating a connection environment variable and having it on the environment part of my `docker-compose.yml` file.

**Reply by Taragolis** [reply]
>The main problem I was having was that if you do airflow connections get NAME_OF_CONNECTION it will give you back a connection URI which was not correct due to the password being a wrong one (is similar but not the same). Is this intended? Is this an issue?

Could you provide information which Airflow version do you use? There was issue with this command in the past, and should be fixed in 2.5.2
- https://github.com/apache/airflow/issues/28766
- https://github.com/apache/airflow/pull/28922

**Reply by ggsdc** [reply]
I am using airflow 2.7.1

**Reply by Taragolis** [reply]
I think it was something wrong with fields, I create connection with password field `bar` (according to the code it use as API Key)

![image](https://github.com/apache/airflow/assets/3998685/caebd4f6-6d2a-4359-9132-4fb51ae17eb2)

And `airflow connections get` returns valid connection URI

```shell
❯ airflow connections get sample_conn
id  | conn_id     | conn_type | description       | host              | schema | login | password | port | is_encrypted | is_extra_encrypted | extra_dejson | get_uri                              
====+=============+===========+===================+===================+========+=======+==========+======+==============+====================+==============+======================================
117 | sample_conn | email     | Sample Connection | smpt.sendgrid.net |        | foo   | bar      | 587  | True         | True               | {}           | email://foo:bar@smpt.sendgrid.net:587
```

```python
from airflow.models.connection import Connection
conn = Connection(uri="email://foo:bar@smpt.sendgrid.net:587")
assert conn.password == "bar"
```

You could also check it with actual API Key locally, maybe it is some issues with encoding some characters. Or it also possible that your password contains whitespaces in the beginning or in the end

**Comment by ggsdc** [comment]
I close the discussion
